doc/cfg_reload_xact_enhance/Config_Reload_Enhancement_Xact_HLD.md

```diff
@@ -0,0 +1,230 @@
+- [Architecture Design](#architecture-design)
+- [High-Level Design](#high-level-design)
+  * [The Temporarily Files Generated by This Enhancement](#the-temporarily-files-generated-by-this-enhancement)
+  * [New Error Messages by This Enhancement](#new-error-messages-by-this-enhancement)
+  * [CLI/YANG model Enhancements](#cli-yang-model-enhancements)
+  * [Config DB Enhancements](#config-db-enhancements)
+- [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+- [Restrictions/Limitations](#restrictions-limitations)
+- [Testing Requirements/Design](#testing-requirements-design)
+  * [Unit Test Cases](#unit-test-cases)
+  * [System Test Cases](#system-test-cases)
+- [Open/Action Items](#open-action-items)
+
+### Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 2022/3/18   |     Charlie Chen   | Initial version                   |
+
+### Scope
+
+This document describes the enhancement based on the existing design of the
+Sonic CLI command 'config reload'. The enhancement introduces the transaction
+mechanism to guarantee that only one 'config reload' session runs in the system
+at the same time. The design regarding the determination of a transaction done
+is included. The design related to how to do 'config reload' is out of the scope.
+
+### Definitions/Abbreviations
+
+| Definition                    | Description                                                                                                          |
+|-------------------------------|----------------------------------------------------------------------------------------------------------------------|
+| Config Reload Process         | A process which is resposnible to execute the config reload operation. Abbreviated to 'CFG RELOAD PROC'.             |
+| Config Reload Monitor Process | A daemon process which will be forked at the start of a config reload session. Abbreviated to 'MONITOR RELOAD PROC'. |
+
+### Overview
+
+The current design of the CLI command 'config reload' is like a simple oneshot command. It does not have any mechanism to prevent from the situations that a 'config reload' session is issuing from the different login session when there is already an unfinished 'config reload' session on the same device. It takes more time to let a device reach an operational state to start another 'config reload' when the previous one has not finished yet. Besides, the successful return of the CLI command on the current design only implies that the operations which are required to initiate the reloading of the configurations are executed without error. This enhancement will fork a new process together with the execution of 'config reload' to monitor the progress of the operation including the state of the critical sonic fundamental services and records down the operation result at the end of the transaction.
```

Comment:
As discussed in today's meeting, we do have checks in place to prevent the parallel config reload operation i.e when the system is not up we don't allow another 'config reload' to go through, please state in which scenario you encountered the issue exactly.

https://github.com/Azure/sonic-utilities/blob/master/config/main.py#L1358

Reply:
I have checked the commit history and find the code you mentioned are added by the PR listed below.

1. https://github.com/Azure/sonic-utilities/pull/1664
2. https://github.com/Azure/sonic-utilities/pull/1766

PR #1664 can prevent the parallel config reload operations under the condition added by that commit. The intention of the PR is not the same as this enhancement. The intention of this enhancement is not only to avoid parallel operations but also to add more information on the processing status of the config reload operation and the final result of the operation. "Prevent parallel config reload operations" is one of the function brings in by this enhancement. Please allow me to have some time to rewrite the HLD to express my idea more clearly.

Reply:
After re-visiting the error logs of sonic-mgmt test cases which are related to the failures of 'config reload' operations in our internal reports, they can be categorized into two types.

**Type 1: The config reload executes on the time that the device just boots up in a short time** 
In sonic-mgmt pytest 'test_vrf.py', it performs `reboot` in the function [`restore_config_db()`](https://github.com/Azure/sonic-mgmt/blob/master/tests/vrf/test_vrf.py#L399). In our pytest environment, a 'config reload' operation will be executed between the two pytest sessions. Upon the end of 'test_vrf.py' session, the execution of 'config reload' gets failed. [sonic-utilities PR #1664](https://github.com/Azure/sonic-utilities/pull/1664) already solves this problem.

**Type 2: The config reload executes on the time that the syncd crashed**
In the termination of a sonic-mgmt pytest session, 'syncd' is crashed due to some issues. Execute `config reload` at that time causes the failure similar to the issue described in [Azure/sonic-utilities #1814](https://github.com/Azure/sonic-utilities/pull/1814). According to the content of the syslog when the issue happens, the failure of 'syncd' triggers the restart of 'swss' through the service auto restart mechanism. The systemctl operations from the service auto restart and from `config reload` are conflicted against `swss` and results in the similar condition described in #1814. This issue is resolved in #1814.

In summary, the proposed mechanism in this HLD does not solve the above two issues actually. I thought the HLD might help on the `config reload` issues that I've ever heard on the 'sonic-mgmt' test cases before I look into it in details. Sorry for the confusion. I will remove the paragraph regarding the issue in `sonic-mgmt` tests from the HLD and add more functionalities and description on the part directly related to "transaction concept". Thanks for your insightful comment to push me clarifying those issues.